### DISCUSSION TITLE
rand: Roadmap for Phase 3 of Reorganize Random

### DISCUSSION THREAD

**Discussion by hungrybluedev** [vlang/v#11794] [Ideas]
## Introduction

This issue is a roadmap for Phase 3 and community contributions are welcome. It is a continuation of the work I, @Delta456 and @x0r19x91 had done last summer. It aims to add support for non-uniform probability distributions (like Gaussian, Poisson, etc) and utility functions (like `shuffle`, `perm`, etc).

There exists some overlap with the existing (and ongoing) work with [VSL](https://github.com/vlang/vsl), specifically the combinatorics module. Hence, discussion is necessary. Please add your suggestions as comments in this issue.

### Move Seeding Functions to `rand.seed`

- [x] Rename `rand.util` to `rand.seed`
- [x] Rename `rand.util.util.v` to `rand.seed.seed.v`

### Distributions Submodule (`rand.dist`)

- [x] ~Functions will support user provided RNGs through sum types and configuration structs.~ Users can change the default RNG to suit their needs. (https://github.com/vlang/v/pull/9083)
- [x] Bernoulli Trial
- [x] Binomial Trial
- [x] Gaussian (or Normal) Distribution Functions
- [x] Exponential Distribution

### Utility Functions (`rand.util`)

- [x] Generic array `shuffle` function.
- [x] Sampling functions

### Final Touches
- [x] Code Review
- [x] Update Documentation

### Postponed (good PR ideas)

- [ ] Poisson Distribution
- [ ] Weibull Distribution
- [ ] Log-normal Distribution
- [ ] Pareto Distribution
- [ ] Beta, Gamma  Distributions
- [ ] von Mises Distribution
- [ ] Triangular Distribution
- [ ] Random array generation tutorial using sampling functions
- [ ] Others...

**Comment by hungrybluedev** [comment]
~A lot of the features I mentioned already exist in @Delta456's [random](https://github.com/Delta456/random) module. We could update it and have it transferred to the V organization.~

After some discussion, it was decided to keep this work in the `rand` module and create two new submodules:
1. `rand.dist` - containing functions related to various probability distributions
2. `rand.util` - containing functions related to shuffling, sampling, and generation of arrays and other collections

**Comment by spytheman** [comment]
Feature request: please add this utility function or equivalent:
```v
// byte returns a uniformly distributed pseudorandom 8-bit unsigned positive `byte`.
pub fn byte() byte {
    return byte(default_rng.intn(256))
}
```
